### ISSUE TITLE
FRAGMOS - additional Enum value in quantityChange -> directionEnum

### ISSUE THREAD

**Issue by JBZ-Fragmos** [finos/common-domain-model#2557]
**BACKGROUND**

let's recap situation as of today : 

![image](https://github.com/finos/common-domain-model/assets/92978493/0341d5a5-7959-42f1-a41e-7839266a706a)

3 Enum value exist in quantityChange Instruction, for the purpose of being used by implementors as a trigger for the distinct functional effects that a quantiyChnage may have on the corresponding PriceQuantity objects in the TradeLot :

- **INCREASE and DECREASE** are describing an **incremental change** which functional meaning is _**straightforward**_ :
   - either positive increment when INCREASE that means an addition -> **new value = old + change**
   - else negative increment when DECREASE that means a substraction -> **new value = old - change**

- **REPLACE** is obvioulsy describing a **replacement change** yet related functional meaning is _**not fully clear**_
   - actually it is ambiguous, because it could cover two distinct types of expected change  -> **new value = ? ? ?** 
      hence a _problem statement_, as described explained below.

**PROBLEM STATEMENT**

Issue is that the REPLACE value may potentially refer to _**two types**_ of replacement - that needs clarifications :
- _either_ expected effect is to **only update a sub-set of  PriceQuantity components** without impacting other prior existing components inside TradeLot
- _else_ it is to **overwrite all prior existing PriceQuantity components** that is to say effect is actually equi valent to **"Cancel And Replace All"** meaning that the whole TradeLot is entirely cancelled and replaced by the new version

None of the two types is preferable per se, either may be required, depending business event intent, etc. also considering internal booking constraints of legacy systems that implementors want to map with CDM.

Accordingly, **both 2 types are complementary**, hence we need to distinguish between them when quantityChange is used, for the purpose of determining clear trigger for implementors to code the relevant functional effect - hence the proposal below.

**PROPOSAL**

Remove existing REPLACE and have **2 new Enum values** instead : 

- **REPLACE_MATCHING_COMPONENTS** to designate first type
- **CANCEL_AND_REPLACE_ALL** to designate second type

![image](https://github.com/finos/common-domain-model/assets/92978493/f87b06e8-a2c4-4b4b-8a7d-94d036eb74a9)

**Comment by rvincentISDA** [comment]
Is this not already dealt with by the function Create_QuantityChange and in particular the alias newPriceQuantity that calls on the function UpdateAmountForEachMatchingQuantity? The latter will decrease or replace all the matching quantity depending on the direction passed as an input.  
I do note though that the UpdateAmountForEachMatchingQuantity does not have any body. I guess that means the implementer ought to provide the implementation ?

**Comment by rvincentISDA** [comment]
![image](https://github.com/finos/common-domain-model/assets/66018066/c816a28f-6527-4831-be8d-9802bb38fb31)

**Comment by JBZ-Fragmos** [comment]
**3 considerations** in regards of your last comment @rvincentISDA that maybe such functional behavior for replacing quantityChange components might already exist :

- yes it exits and works but on very **restricted scope** of business case : actully the current function - which indeed is supposed to have a behavior aligned with the Enum values i've described for convenience as "REPLACE_MATCHING_COMPONENTS" when value is "REPLACE" - does NOT work properly whenever the set of PriceQuantity(*) objects is more complex than only single PQ changed... in one word, current functions works for very vanilla cases, and is failing to cover mostly all other cases e.g. instead of matching it creates duplicates, etc.

    - FYI, we have already developped our own functions in Fragmos-Chain for the purpose of handling any business cases involving any TradeLot(*)->PriceQuantity(*) arrays
    
- the current function with "REPLACE" also does not have **"additive effect"** as we propose above i.e. we are proposing that the functional effect shall be : when existing PQ is matched, replace it, else when PQ in quantityChange is new (does not exist yet in prior TradeLot.PQ) then such new PQ is added (= replaces "nothing" by "something" if we may say...) : without such functional effect, it is impossible to update existing PQ, say for instance an existing PQ ha only 1 price, and you want to add 2nd price for instance in other currency, then simple matching as of today does not match anything, so you cannot do it... hence need to have adding effect for such new unmatchable price

-  the current function with "REPLACE" obviously is **one kind only among two possible kinds** of replacement : it is of kind "REPLACE_MATCHING_COMPONENTS" and one purpose of this Issue is notably to distinguish / add complementary kind of replacement that is "CANCEL_AND_REPLACE_ALL"

**Comment by hugohills-regnosys** [comment]
> Is this not already dealt with by the function Create_QuantityChange and in particular the alias newPriceQuantity that calls on the function UpdateAmountForEachMatchingQuantity? The latter will decrease or replace all the matching quantity depending on the direction passed as an input.
> I do note though that the UpdateAmountForEachMatchingQuantity does not have any body. I guess that means the implementer ought to provide the implementation ?

There's a Java implementation, `UpdateAmountForEachMatchingQuantityImpl`

**Comment by JBZ-Fragmos** [comment]
thanks for feedback, 

yet let me rephrase then as below, for clarity :

- **main purpose** of this Issue is to release an **additional new type** of replacement trigger which for sure does **not** already exist - this is the one i have named "CANCEL_AND_REPLACE_ALL"
      - **and then** accordingly I also propose to **rename existing one** "REPLACE" just for distinguishing it with new type, by "REPLACE_MATCHING_COMPONENTS"

- besides, as regards current one REPLACE (whether or not renamed as REPLACE_MATCHING_COMPONENTS) i would also suggest that functional effect shall be modified i.e. as of today say you have existing PQ1 and PQ2 and in change you have fully new PQ3 you cannot match neither with PQ1 nor PQ2 yet you would like to add this PQ3, then you would need REPLACE behavior to be "martly additive" meaning that when not matching PQ3, then such PQ3 would be added to existing ones, thus as a result of your change you would have PQ1, PQ2 and PQ3 present... 
     - which is not same need as for instance to replace both PQ1 and PQ2 by PQ3 in order to get as a result only one PQ being PQ3 and both prior ones removed, hence the 2nd type of replacement trigger for such other type of need

**Comment by lolabeis** [comment]
@tomhealey-icma This may bear some relationship with issue #2581 that you just raised

**Comment by tomhealey-icma** [comment]
I've experienced the same problem and I would support this change.

While this will fix the problem it limits the ability to make targeted updates to specific price/quantities. One possibility would be, since the model already uses tradeLot identifiers, why not add priceQuantity identifiers?  It may make sense to review the relationship of tradeLot prices/quantities to other parts of the model specifically the product. Far more complicated scenarios are emerging without a clear and transparent relationship between tradeLots and other parts of the model.

**Comment by JBZ-Fragmos** [comment]
@tomhealey-icma, yes, but not done yet, i will create the PR once main points being discussed are clarified
@lolabeis will further document and answer your question by the end of this week - thks

**Comment by JBZ-Fragmos** [comment]
@tomhealey-icma juts one remeark as regards your comments about **TradeLots** and related **referencing**, notably inside Product, etc i fully agree there are some problem statements about it, yet should be discussed as part of **another Issue**... Therefore have created a new one to discuss this point, please see https://github.com/finos/common-domain-model/issues/2582#:~:text=FRAGMOS%20%2D%20Referencing%20TradeLots%20PQs%20in%20Product%20%232582

**Comment by JBZ-Fragmos** [comment]
@lolabeis 

- **use case for failing REPLACE** : 
   - for 1 for 1 replacement that would fail, sorry not sure i can reproduce easily cases where REPLACE might fail because we have implemented our own functions since quite a time now, so would have to re-plug , back to Rosetta functions previously used for reproducing, and have little time to coordinate such request with my Dev teams these days... anyway it seems @tomhealey-icma had some recent use cases to provide, since looks like similar issue occurred in https://github.com/finos/common-domain-model/issues/2581
   - for specific case I had in mind, please consider this situation : 
     - given PQ for observable "StockA" has ony one Q, say Q1 that is 20 financialUnit, Share... so initial situation : PQ=Q1
     - and now you decide to update this information by adding as well another Q, say Q2 that could be related nominal amount, say 200 currency,EUR... meanin at target you want to have updated PQ=Q1,Q2
     - i guess the REPLACE function cannot match existing PQ=Q1 with change PQ=Q1;Q2, therefore event builder will fail... 
     - this need for the function to add unmatched P or unmatched Q to existing PQ for which observable is matched, is what i'm refering to in my description as "smart additive" behavior
    
- **use-case for the REPLACE_ALL scenario** happens mostly when there is a need to remove given PQ, or to add new P or new Q to an existing PQ... could be done by matching components with "cancelling" effect, yet i'm more in favor of cancel/replace the whole in such cases, because mentioning in event message the component that shall be removed is tricky for some legacy systems once such item is removed or overwritten by the new one, then mapping is quite difficult to perform, whereas taking the whole new version for mapping the fact previous version has been fully replaced is easier :
     - **case 1 = remove P or Q** : reverse example above : say you have PQ=Q1,Q2 and for some reason you would like to have only PQ=Q1... meaning you want to replace something (Q2) by nothing because you want to remove Q2... current REPLACE function cannot make it, hence CANCEL_REPLACE should complete cureetn function for such need
     - **case 2 = remove Observable** : you need to remove 1 or several observable... PQ=StockA shall be fully removed, for some reason, for instance because of a delisting event, or because traders decided to substitute, say maybe PQ=StockA will be replaced by PQ=StockB,StockC, etc. in all these cases, there is a need to **cancel** PQ=StockA and maybe also to replace it by, say PQ=StockB, Stock...

**Comment by JBZ-Fragmos** [comment]
@lolabeis also I would like to clarify that maybe we shall distinct two types of items : 
- type = "object" : original request only concern this item i.e. an Enum value, nothing else
- and type = "func" : original request is not expecting functional rules to be fully implemented I was just describing such functional rules for the purpose of providing context why asking for 4th Enum value to be released 

what I mean is that all in all this Story had 1 main purpose, that was to ask for **adding 4th Enum value in existing list**, that measn to update the type object "directionEnum", so that instead of INCREASE, DECREASE, REPLACE, we may have also REPLACE "ALL", and for the purpose of clarifying why and also avoid confusion with existing REPLACE i've prosposed all in all to have as below 
- INCREASE (no change)
- DECREASE (no change)
- REPLACE is just renamed as REPLACE_MATCHING_COMPONENTS
- CANCEL_AND_REPLACE_ALL is a new one we add to the list

I'm trying to clarify the fact that modifying Enum value list shall be distinguished from defining all the related funcitonal rules that might use such Enum values... There are 2 distinct steps : 
- discuss/agree/release Enum list of value once related functional meaning for being "triggers" is clarified, that is purpose of current request i.e. considring directionEnum as a "trigger" list I'm proposing 1 new trigger ("CANCEL_AND_REPLACE_ALL")
- then discuss/agree/release "func" pieces of codes which effectiveley implement and use such "tiggers"..; that might required separate discussion but shall not block implementors
